apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/FieldsAdapter.java

```diff
@@ -0,0 +1,71 @@
+package com.apollographql.android.cache.normalized;
+
+import com.apollographql.android.impl.BufferedSourceJsonReader;
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.math.BigDecimal;
+import java.util.Map;
+
+import okio.BufferedSource;
+import okio.Okio;
+
+public final class FieldsAdapter {
+  private final JsonAdapter<Map<String, Object>> serializationAdapter;
+
+  private FieldsAdapter(Moshi moshi) {
+    Type type = Types.newParameterizedType(Map.class, String.class, Object.class);
+    serializationAdapter = moshi.adapter(type);
+  }
+
+  public static FieldsAdapter create() {
+    Moshi moshi = new Moshi.Builder()
+        .add(CacheReference.class, new CacheReferenceAdapter())
+        .add(BigDecimal.class, new BigDecimalAdapter())
+        .build();
+    return new FieldsAdapter(moshi);
+  }
+
+  public String toJson(Map<String, Object> fields) {
+    return serializationAdapter.toJson(fields);
+  }
+
+  public Map<String, Object> fromBufferedSource(BufferedSource bufferedSource) throws IOException {
+    CacheJsonStreamReader jsonStreamReader
+        = new CacheJsonStreamReader(new BufferedSourceJsonReader(bufferedSource));
```

Comment:
Personally I like APIs with static calls for object creation like 
`ApolloReader.cacheJsonStream(ApolloReader. bufferedSourceJson(bufferedSource)`
This way you can do static imports and become `cacheJsonStream(bufferedSourceToJson(source))`